### DISCUSSION TITLE
Compatibility with Microsoft Entra multifactor authentication external method provider

### DISCUSSION THREAD

**Discussion by nicolaspernoud** [authelia/authelia#8431] [Q&A] [answered]
Hello, and thanks for this product !

Is it compatible with "Microsoft Entra multifactor authentication external method provider" (https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider) ?
The use case is to delegate the checking of another factor to the Idp, while using Entra Id as main IdP.

For this to work, there would be a few requirements :
- implement the implicit flow, or at least the `id_token` response type :  https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider#microsoft-entra-id-call-to-the-external-identity-provider ;
- parse the id_token_hint provided by entra id : https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider#example-of-an-id_token_hint ;
- provide an id_token that must be an mfa token : https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider#example-of-an-eam-that-satisfies-mfa ;
- be extensible to integrate custom authentication methods.

**Comment by james-d-elliott** [comment, marked as answer]
We do not currently, we support the OpenID Connect 1.0 and many related standards within the context of the provider role, inclusive of all of the standardized flows with the exclusion of deprecated flows and the device code flow which is coming in the next version, as well as future support for the relying party role in later versions.

Also unfortunately based on my reading of the document (which is also unfortunately very vague and unclear in my opinion, as in part evidenced by `provide an id_token that must be an mfa token` which seems to actually be a [authorization endpoint claims parameter](https://openid.net/specs/openid-connect-core-1_0.html#ClaimsParameter) not token contents) this is unlikely to get a whole lot of traction in most projects including this one.

Some of the reasons for this that I've identified briefly are as follows:

1. Microsoft's "OpenID Connect 1.0" implementation blatantly ignores/breaks **_MUST_** requirements from the specification broadly speaking in multiple ways.
2. This specific extension:
    1. Is not a formal standard.
    2. Is very abstractly defined.
    3. Uses a less secure flow than the standard authorization code flow.
    4. Could be implemented purely through standardized flows like:
        1. The base specification using the Authorization Code flow leveraging the existing ACR and AMR claims and associated authorization parameters.
        2. The OAuth 2.0 Token Exchange flow.
3. Introducing complex handling logic into an already complex area is not very wise for projects wanting to avoid serious security vulnerabilities, especially when adding the logic is effectively hijacking well established flow logic from the standards and introducing a completely different way to handle it, which is as far as I can tell what's occurring here; though their docs on this feature are incredibly lacking.
4. The OAuth 2.0 ecosystem is already very complicated, these parameters are used by various other mechanisms to perform very specific functionality; this in my opinion makes it incredibly irresponsible for implementers (such as Microsoft in this example) to be implementing extensions outside of the specs without:
    1. Explaining the reason the standard flows do not accomplish the desired goal.
    2. Engaging the OpenID Connect 1.0 working group to develop a standard surrounding this so that implementers know the details are stable and not subject to change upon the whims of a single company.

**Comment by nicolaspernoud** [comment]
Thanks for your reply.
I would seems that Microsoft is doing here what it has always done : trying to break the standard by sheer market weight...
It is then understandable that products do not follow blindly, even if it would be useful for users...

**Reply by james-d-elliott** [reply]
Yeah the STS they're talking about is precisely what Token Exchange was designed to handle, it is very strange what they're doing. Usage of the `id_token_hint` parameter in the authorization endpoint is solely for tokens issued by the authorization server or authorization server cluster that issued the token, this is very apparent when reading the following sections:

- https://openid.net/specs/openid-connect-core-1_0.html#AuthRequestValidation
- https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest

Using it for other purposes like they are especially in a multi-tenancy environment especially when the standard validation process for this parameter requires you ignore certain critical standard validation checks (expiration etc) doesn't seem like a great idea.
